### ISSUE TITLE
Shadow DOM emulation does not correctly retarget events.

### ISSUE THREAD

**Issue by jelbourn** [angular/angular#1168]
The Shadow DOM spec details [event retargeting](http://www.w3.org/TR/shadow-dom/#event-retargeting). This basically means that when an event propagation hits a shadow root boundary, it keeps propagating, but the event target is changed to the host element. 

It's also important to note that [some event types are never propagated outside of a shadow root](http://www.w3.org/TR/shadow-dom/#events-that-are-always-stopped).

Angular's shadow DOM emulation currently doesn't do this, and it makes it confusing to try to add listeners to events like `click` and have to use the `^click` syntax (see #1040).

**Comment by naomiblack** [comment]
@jelbourn should this be M7?

**Comment by jelbourn** [comment]
If GreenTea is using Native shadow DOM mode, then it wouldn't affect them. For me, it's easy to workaround.

**Comment by tbosch** [comment]
Blocked on design for our custom reprojection.

**Comment by jelbourn** [comment]
@tbosch is this still blocked?

**Comment by tbosch** [comment]
No,
it's not. We now have our own reprojection in place, and are free to decide whether we actually want to retarget events at all, as we are using a custom tag (`<ng-content>`) which does not interfere with the native `<content>` element.

**Comment by jelbourn** [comment]
@tbosch Retargeting is a must, even if we aren't using the native `content`. Without it, implementation details are leaked outside of a component's API surface area.

**Comment by tbosch** [comment]
So how did it work in Angular1?

**Comment by jelbourn** [comment]
@tbosch AngularJS just lets all events propagate and fire handlers normally. E.g., if you put `ng-click="doSomething()"` on an element, `doSomething` will be called any time a descendant element is clicked. The is the same behavior as `onclick` or `addEventListener`. 

I know little of Angular.dart, but I suspect it does the same thing. @yjbanov could confirm. 

Angular 2 is different because it has the default behavior of only invoking event handlers if the event target matches the handling element. If this behavior continues to exist without event retargeting, users end up in a weird situation where, without native shadow DOM, they have to know whether a component's event is coming from the component directly, or is being propagated from some internal element.

**Comment by naomiblack** [comment]
@tbosch now that content reprojection, is this unblocked? @jelbourn is open to working on this with some direction. Can you direct in alpha-35...?

**Comment by tbosch** [comment]
Almost unblocked. I am reworking ShadowDomStrategy right now, but should be done in 1-2 days (see https://github.com/angular/angular/pull/3314).

**Comment by jelbourn** [comment]
@tbosch Is the unblocked now?

**Comment by tbosch** [comment]
Yes, unblocked.

**Comment by mhevery** [comment]
Now that we allow DOM events to bubble. Is this still an issue?

What ShadowDOM re-targeting does is that it changes the `target` property of the event as if the event come form the component to hide the shadow DOM as an implementation detail. Not sure it is that important.

@jelbourn can you come up with a concrete example where we would need this? I think having such an example would make it easier to discuss what is needed.

**Comment by jelbourn** [comment]
This is less of an issue now that DOM events can bubble, but I think that event retargeting should still happen. Components should also probably have some control over whether they want to retarget or not. 

The use-case here is any Component that wants to act like a widget / form control. Say, for example, you have a Material Design button. Its rendered output would look something like
```html
<button md-button>
  <div md-ripple-container>
    Save
  </div>
</button>
```

As far as the consumer of the `MdButton` is concerned, the target of a click event here should be the `button` element that they put on the page. Without retargeting, though, the target is going to be the `md-ripple-container` div, which is leaking implementation details of the component.

**Comment by mhevery** [comment]
This is obsolete, and no longer relevant / actionable. To keep our issues clean, we are aggressively closing them.
